sympy/functions/combinatorial/factorials.py

```diff
@@ -1058,31 +1066,34 @@ def _eval_expand_func(self, **hints):
 
     def _eval_rewrite_as_Piecewise(self, n, k, **kwargs):
         from sympy.concrete.products import Product
-        from sympy import gamma
+        from sympy import gamma, Abs, nan
         knk = lambda x: Or(Eq(k, x), Eq(n - k, x))
         isint = lambda x: Eq(x, floor(x))
         nint = lambda x: Ne(x, floor(x))
-        i = Dummy('i', integer=True, positive=True)
+        pos = lambda x: And(Ne(x,0), Eq(x/Abs(x), 1))
+        neg = lambda x: And(Ne(x,0), Eq(x/Abs(x), -1))
+        fin = lambda x: Eq(x/x, nan)
+        for i in igen():
+            if not n.has(i) and not k.has(i):
+                break
         return Piecewise(
-            (1, knk(1)),
+            (1, Or(Eq(n, k), knk(0))),
+            (n, knk(1)),
+            (zoo, And(neg(n), isint(n), nint(k))),
+            (nan, Or(fin(n), fin(k))),
             (0, Or(
-                knk(0),
+                knk(-k),
```

Comment:
`knk(-k)` is equivalent to `n =0 or k = 0` and `binomial(3,0) is 1, not 0`

Reply:
```python3
 (1, Or(Eq(n, k), knk(0))),
```
This precedes that statement right

Reply:
Yes...further proof that it doesn't belong in the 0 result predicate or that it should be changed to something else if it is needed.

Reply:
Our 0 conditions are:
```
        if -1 in (k, nk) and (n.is_integer is False or (n + 1).is_zero is False):
            return S.Zero
        if all(x.is_integer for x in (n, k)):
            if (n*k).is_positive and (k - n).is_positive:
                return S.Zero
            if n.is_nonnegative and k.is_negative:
                return S.Zero
```
So I don' think the `knk(-k)` should be there.

Reply:
Alright, I'll change that 👍